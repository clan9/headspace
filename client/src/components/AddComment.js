import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter, Link } from 'react-router-dom';
import moment from 'moment';
import { addComment } from 'actions/posts';

export class AddComment extends Component {
  state = { text: '', createdAt: moment(), error: '' };

  onChange = e => {
    const text = e.target.value;
    this.setState(() => ({ text }));
  };

  resetError = () => {
    setTimeout(() => {
      this.setState(() => ({ error: '' }));
    }, 6000);
  };

  componentWillUnmount() {
    this.setState(() => ({ error: '' }))
  }

  onSubmit = e => {
    e.preventDefault();

    if (!this.state.text) {
      this.setState(() => ({
        error: 'You cannot add an empty comment'
      }));
      this.resetError();
    } else {
      this.setState(() => ({ error: '' }));
      this.props.addComment({
        text: this.state.text,
        createdAt: this.state.createdAt.valueOf(),
        id: this.props.id,
        history: this.props.history
      });
    }
  };

  render() {
    return (
      <div className="content-container">
        <h2 className="page__heading">Add a Comment</h2>
        <form onSubmit={this.onSubmit} className="form">
          <div className="form__error__container">
            {this.state.error && (
              <p className="form__error">{this.state.error}</p>
            )}
          </div>

          <textarea
            type="text"
            placeholder="What did you think?"
            className="form__textarea"
            value={this.state.text}
            onChange={this.onChange}
          />
          <div className="comment__form__button__container">
            <Link to="/dashboard" className="comment__form__button">
              Home
            </Link>
            <button className="comment__form__button" type="submit">
              Save Comment
            </button>
          </div>
        </form>
      </div>
    );
  }
}

AddComment.propTypes = {
  addComment: PropTypes.func.isRequired
};

const mapStateToProps = (state, ownProps) => ({
  post: state.posts.post,
  id: ownProps.match.params.id
});

export default connect(
  mapStateToProps,
  { addComment }
)(withRouter(AddComment));
