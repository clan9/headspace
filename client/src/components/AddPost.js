import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { addPost } from 'actions/posts';
import PostForm from 'components/PostForm';

export const AddPost = ({ addPost, history }) => {
  return (
    <div className="content-container">
      <h2 className="page__heading">Add a Post</h2>
      <PostForm
        onSubmit={post => {
          addPost(post, history);
        }}
      />
    </div>
  );
};

AddPost.propTypes = { addPost: PropTypes.func.isRequired };

export default connect(
  null,
  { addPost }
)(withRouter(AddPost));
