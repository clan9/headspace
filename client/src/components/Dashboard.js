import React, { useEffect } from 'react';
import PostList from 'components/PostList';
import PostFilter from 'components/PostFilter';
import { connect } from 'react-redux';

export const Dashboard = ({ name = '' }) => {
  const nameCasing = () => {
    if (name) {
      return name.charAt(0).toUpperCase() + name.slice(1);
    }
  };

  useEffect(() => {
    window.scrollTo(0,0)
  }, [])

  return (
    <div className="content-container">
      <h2 className="page__heading">Welcome {nameCasing()}</h2>
      {!name && (
        <p className="dashboard__text">
          In order to submit a post, comment or like, please sign up for a new
          account or log in.
        </p>
      )}
      <PostFilter />
      <PostList />
    </div>
  );
};

const mapStatetoProps = state => ({
  name: state.profile.name
});

export default connect(mapStatetoProps)(Dashboard);
