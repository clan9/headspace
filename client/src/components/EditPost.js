import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { editPost } from 'actions/posts';
import PostForm from 'components/PostForm';

export const EditPost = ({ post, id, history, editPost }) => {
  return (
    <div className="content-container">
      <h2 className="page__heading">Edit your Post</h2>
      <PostForm
        post={post}
        onSubmit={post => {
          editPost(id, post, history);
        }}
      />
    </div>
  );
};

EditPost.propTypes = {
  editPost: PropTypes.func.isRequired
};

const mapStateToProps = (state, ownProps) => ({
  post: state.posts.post,
  id: ownProps.match.params.id
});

export default connect(
  mapStateToProps,
  { editPost }
)(withRouter(EditPost));
