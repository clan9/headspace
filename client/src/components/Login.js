import React from 'react';
import { Link } from 'react-router-dom';

const Login = props => {
  return (
    <div className="login">
      <div className="login__box">
        <h1 className="login__title">headSpace</h1>
        <p className="login__text">
          Share your thoughts and connect with others!
        </p>
        <Link to="/signup" className="login__buttons cta">
          Sign Up!
        </Link>
        <Link to="/signin" className="login__buttons">
          Log In
        </Link>
        <Link to="/dashboard" className="login__buttons">
          Browse Posts
        </Link>
      </div>
    </div>
  );
};

export default Login;
