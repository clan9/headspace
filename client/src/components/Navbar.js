import React from "react";
import PropTypes from "prop-types";
import { NavLink } from "react-router-dom";
import { connect } from "react-redux";
import { logout } from "actions/auth";

export const Navbar = ({ logout, auth: { isAuthenticated, loading } }) => {
  const userLogout = () => {
    logout();
  };

  const renderLinks = () => {
    if (!isAuthenticated && !loading) {
      return (
        <div className="navbar">
          <div className="navbar__left">
            <NavLink
              to="/dashboard"
              activeClassName="is-active"
              className="navbar__link navbar__link--logo"
            >
              headSpace
            </NavLink>
          </div>

          <div className="navbar__right">
            <NavLink
              to="/signup"
              activeClassName="is-active"
              className="navbar__link"
            >
              Sign Up
            </NavLink>
            <NavLink
              to="/signin"
              activeClassName="is-active"
              className="navbar__link"
            >
              Log In
            </NavLink>
          </div>
        </div>
      );
    } else {
      return (
        <div className="navbar">
          <div className="navbar__left">
            <NavLink
              to="/dashboard"
              activeClassName="is-active"
              className="navbar__link navbar__link--logo"
            >
              headSpace
            </NavLink>
          </div>
          <div className="navbar__right">
            <NavLink
              to="/profile"
              activeClassName="is-active"
              className="navbar__link"
            >
              Your Stats
            </NavLink>
            <NavLink to="/" onClick={userLogout} className="navbar__link">
              Logout
            </NavLink>
          </div>
        </div>
      );
    }
  };

  return <div>{renderLinks()}</div>;
};

Navbar.propTypes = {
  logout: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth
});

export default connect(
  mapStateToProps,
  { logout }
)(Navbar);
