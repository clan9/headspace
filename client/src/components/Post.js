import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import moment from 'moment';
import { fetchPost, deletePost, likePost, unlikePost } from 'actions/posts';

export const Post = ({
  fetchPost,
  deletePost,
  post,
  comments,
  id,
  loading,
  userId,
  author,
  history,
  likePost,
  unlikePost,
  error
}) => {
  useEffect(() => {
    fetchPost(id);
  }, [fetchPost, id]);

  const renderComments = () => {
    if (post.comments) {
      return post.comments.map(comment => (
        <div key={comment._id} className="post__comments__item">
          <p className="post__comments__item__text">
            <span className="post__comments__item__text--name">
              {comment.name} said:{' '}
            </span>
            {comment.text}
          </p>
          <p className="post__comments__item__time">
            {moment(comment.createdAt).fromNow()}
          </p>
        </div>
      ));
    }
  };

  const renderActions = () => {
    if (userId && userId === author) {
      return (
        <div className="post__actions">
          <Link className="post__actions__item" to={`/edit/${id}`}>
            Edit your post
          </Link>
          <Link className="post__actions__item" to={`/comment/${id}`}>
            Add a Comment
          </Link>
          <button
            className="post__actions__item"
            onClick={() => onDelete(id, history)}
          >
            Delete this Post
          </button>
        </div>
      );
    } else if (userId) {
      return (
        <div className="post__actions">
          <Link className="post__actions__item" to={`/comment/${id}`}>
            Add a Comment
          </Link>
        </div>
      );
    }
  };

  const renderLikeButtons = userId => {
    if (userId) {
      return (
        <div>
          <button className="like-post-button" onClick={onLikePost}>
            Like
          </button>
          <button className="like-post-button" onClick={onUnlikePost}>
            Unlike
          </button>
        </div>
      );
    }
  };

  const onDelete = () => {
    deletePost(id, history);
  };

  const onLikePost = () => {
    likePost(id);
  };

  const onUnlikePost = () => {
    unlikePost(id);
  };

  return (
    <div>
      {loading ? (
        <div>loading...</div>
      ) : (
        <div className="content-container">
          <div className="post">
            <div className="post__heading">
              <h2 className="post__heading--title">{post.title}</h2>
              <p className="post__heading--subtitle">
                by {post.name} on {moment(post.createdAt).format('Do MMM YYYY')}
              </p>
            </div>
            <div className="post__content">
              <p>{post.content}</p>
            </div>
            <div className="post__stats">
              <p>
                Latest edit/comment:{' '}
                <span className="post__stats__info">
                  {moment(post.editedAt).fromNow()}
                </span>{' '}
              </p>
              <div>{renderLikeButtons(userId)}</div>
            </div>
          </div>
          <div className="post__error__container">
            {error && <p className="post__error">{error}</p>}
          </div>
          <div>{renderActions()}</div>
          {comments && comments.length > 0 && (
            <h3 className="post__comments__header">Comments</h3>
          )}
          <div className="post__comments">{renderComments()}</div>
        </div>
      )}
    </div>
  );
};

Post.propTypes = {
  fetchPost: PropTypes.func.isRequired,
  deletePost: PropTypes.func.isRequired,
  loading: PropTypes.bool
};

const mapStateToProps = (state, ownProps) => ({
  post: state.posts.post,
  comments: state.posts.post.comments,
  loading: state.posts.loading,
  id: ownProps.match.params.id,
  userId: state.auth.user._id,
  author: state.posts.post.user,
  error: state.posts.error
});

export default connect(
  mapStateToProps,
  { fetchPost, deletePost, likePost, unlikePost }
)(withRouter(Post));
