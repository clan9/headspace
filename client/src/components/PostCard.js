import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const PostCard = ({
  _id,
  title,
  content,
  likes,
  comments,
  name,
  createdAt
}) => {
  const renderSummary = () => {
    const postSummary = content
      .split(' ')
      .slice(0, 20)
      .join(' ');
    return postSummary;
  };

  return (
    <Link to={`/post/${_id}`} className="post-card__link">
      <div className="post-card">
        <h3 className="post-card__title">{title}</h3>
        <p className="post-card__summary">
          {renderSummary()}.....
          <span className="post-card__summary--more"> more</span>
        </p>
        <div className="post-card__stats">
          <div className="post-card__stats__left">
            <p className="post-card__stats--by">
              By: {name} on {moment(createdAt).format('Do MMM YYYY - HH:mm')}
            </p>
          </div>
          <div className="post-card__stats__right">
            <p className="post-card__stats__right--comments">
              Comments: {comments.length}
            </p>
            <p className="post-card__stats__right--likes">
              Likes: {likes.length}
            </p>
          </div>
        </div>
      </div>
    </Link>
  );
};

export default PostCard;
