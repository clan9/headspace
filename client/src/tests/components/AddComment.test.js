import React from 'react';
import { shallow } from 'enzyme';
import { AddComment } from 'components/AddComment';
import moment from 'moment';

describe('AddComment component tests', () => {
  let addComment, history, id, createdAt, wrapper;

  beforeEach(() => {
    addComment = jest.fn();
    history = jest.fn();
    id = '1';
    createdAt = moment().valueOf();
    wrapper = shallow(
      <AddComment addComment={addComment} id={id} history={history} />
    );
  });

  it('should render AddComment component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render AddComment component correctly when there is an error message to display', () => {
    wrapper.find('form').simulate('submit', { preventDefault: () => {} });
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle changes to the text area', () => {
    const value = 'new text';
    wrapper.find('textarea').simulate('change', { target: { value } });
    expect(wrapper.state('text')).toBe(value);
  });

  it('should handle valid form submission', () => {
    const value = 'new text';
    wrapper.find('textarea').simulate('change', { target: { value } });
    wrapper.find('form').simulate('submit', { preventDefault: () => {} });

    expect(wrapper.state('error')).toBe('');
    expect(addComment).toHaveBeenLastCalledWith({
      text: value,
      createdAt,
      id,
      history
    });
  });

  it('should handle invalid form submission', () => {
    wrapper.find('form').simulate('submit', { preventDefault: () => {} });
    expect(wrapper.state('error')).toBe('You cannot add an empty comment');
    // expect(resetError).toHaveBeenCalled();
  });
});
