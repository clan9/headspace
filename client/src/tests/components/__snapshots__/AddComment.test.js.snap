// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddComment component tests should render AddComment component correctly 1`] = `
<div
  className="content-container"
>
  <h2
    className="page__heading"
  >
    Add a Comment
  </h2>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="form__error__container"
    />
    <textarea
      className="form__textarea"
      onChange={[Function]}
      placeholder="What did you think?"
      type="text"
      value=""
    />
    <div
      className="comment__form__button__container"
    >
      <Link
        className="comment__form__button"
        to="/dashboard"
      >
        Home
      </Link>
      <button
        className="comment__form__button"
        type="submit"
      >
        Save Comment
      </button>
    </div>
  </form>
</div>
`;

exports[`AddComment component tests should render AddComment component correctly when there is an error message to display 1`] = `
<div
  className="content-container"
>
  <h2
    className="page__heading"
  >
    Add a Comment
  </h2>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="form__error__container"
    >
      <p
        className="form__error"
      >
        You cannot add an empty comment
      </p>
    </div>
    <textarea
      className="form__textarea"
      onChange={[Function]}
      placeholder="What did you think?"
      type="text"
      value=""
    />
    <div
      className="comment__form__button__container"
    >
      <Link
        className="comment__form__button"
        to="/dashboard"
      >
        Home
      </Link>
      <button
        className="comment__form__button"
        type="submit"
      >
        Save Comment
      </button>
    </div>
  </form>
</div>
`;
